ZEngine/ZEngine/Core/Maths/Vec.h

```diff
@@ -0,0 +1,266 @@
+        {
+            Derived result;
+            for (size_t i = 0; i < N; ++i)
+            {
+                result[i] = m_data[i] * scalar;
+            }
+            return result;
+        }
+
+        Derived operator/(T scalar) const
+        {
+            Derived result;
+            for (size_t i = 0; i < N; ++i)
+            {
+                result[i] = m_data[i] / scalar;
+            }
+            return result;
+        }
+
+        Derived& operator*=(T scalar)
+        {
+            for (size_t i = 0; i < N; ++i)
+            {
+                m_data[i] *= scalar;
+            }
+            return *this;
+        }
+
+        Derived& operator/=(T scalar)
+        {
+            for (size_t i = 0; i < N; ++i)
+            {
+                m_data[i] /= scalar;
+            }
+            return *this;
+        }
+        Derived operator+(const Derived& other) const
+        {
+            Derived result;
+            for (size_t i = 0; i < N; ++i)
+            {
+                result[i] = m_data[i] + other[i];
+            }
+            return result;
+        }
+
+        Derived operator-(const Derived& other) const
+        {
+            Derived result;
+
+            for (size_t i = 0; i < N; ++i)
+            {
+                result[i] = m_data[i] - other[i];
+            }
+            return result;
+        }
+
+        Derived operator*(const Derived& other) const
+        {
+            Derived result;
+            for (size_t i = 0; i < N; ++i)
+            {
+                result[i] = m_data[i] * other[i];
+            }
+            return result;
+        }
+
+        T dot(const Derived& other) const
+        {
+            T result = T();
+            for (size_t i = 0; i < N; ++i)
+            {
+                result += m_data[i] * other[i];
+            }
+            return result;
+        }
+
+        T magnitude() const
+        {
+            const Derived& self = static_cast<const Derived&>(*this);
+            return std::sqrt(self.dot(self));
+        }
+
+        Derived normalized() const
+        {
+            Derived result;
+            T       len = magnitude();
+            for (size_t i = 0; i < N; ++i)
+            {
+                result[i] = m_data[i] / len;
+            }
+            return result;
+        }
+
+        T cross2d(const Derived& other) const
+        {
+            ZENGINE_VALIDATE_ASSERT(N == 2, "cross2d() is only defined for 2D vectors.");
+            return (m_data[0] * other[1]) - (m_data[1] * other[0]);
+        }
+
+        Derived cross3d(const Derived& other) const
+        {
+            ZENGINE_VALIDATE_ASSERT(N == 3, "cross3d() is only defined for 3D vectors.");
+            Derived result;
+
+            result[0] = m_data[1] * other[2] - m_data[2] * other[1];
+            result[1] = m_data[2] * other[0] - m_data[0] * other[2];
+            result[2] = m_data[0] * other[1] - m_data[1] * other[0];
+            return result;
+        }
```

Comment:
I prefer this Cross op to be defined out of the struct Vec be inlined
eg:
inline Vec3f Cross(const Vec3f& a, const Vec3f& b) {...}